src/relay/pass/call_graph.cc

```diff
@@ -0,0 +1,339 @@
+  CallGraphEntry::CallGraphEntrySet visited;
+
+  for (const auto& it : entries) {
+    // Keep tracking the nodes that have been visited.
+    auto topo = it->TopologicalOrder(&visited);
+    // Preprend the collected items. The intermeidate nodes that are shared by
+    // multiple entries are guaranteed to be collected when visiting the
+    // previous entries. Therefore, topological order remains.
+    ret.insert(ret.begin(), topo.begin(), topo.end());
+  }
+
+  // Find out the missing global functions if there are any to help debugging.
+  if (ret.size() != module->functions.size()) {
+    for (auto it : module->functions) {
+      if (visited.find((*this)[it.first]) == visited.end()) {
+        LOG(WARNING) << "Missing global:" << it.first->name_hint
+                     << " with # refs = " << (*this)[it.first]->GetRefCount();
+      }
+    }
+    LOG(FATAL) << "Expected " << module->functions.size()
+               << " globals, but received "
+               << ret.size();
+  }
+
+  return ret;
+}
+
+// A BSF traverser is used to collect the nodes in a CallGraphEntry. The nodes
+// that are visited by previous CallGraphEntry entries can be memoized. This
+// helps us to make sure no entry will be visited multiple times when collecting
+// the nodes for an entir call graph.
+std::vector<CallGraphEntry*> CallGraphEntry::TopologicalOrder(
+    CallGraphEntrySet* visited) const {
+  std::vector<CallGraphEntry*> ret;
+  std::vector<CallGraphEntry*> current_nodes;
+  if (visited->find(this) == visited->end()) {
+    visited->emplace(this);
+    current_nodes.emplace_back(const_cast<CallGraphEntry*>(this));
+  }
+
+  std::vector<CallGraphEntry*> next_nodes;
+  while (!current_nodes.empty()) {
+    for (const auto& node : current_nodes) {
+      ret.push_back(node);
+      // Iterate through the called entries.
+      for (auto git = node->begin(); git != node->end(); ++git) {
+        if (visited->find(git->second) == visited->end()) {
+          next_nodes.push_back(git->second);
+          visited->emplace(git->second);
+        }
+      }
+    }
+    // Update the current level and clean the next level.
+    current_nodes = next_nodes;
+    next_nodes.clear();
+  }
+  return ret;
+}
+
+void CallGraphEntry::CleanCallGraphEntries() {
+  while (!called_globals_.empty()) {
+    // Decrement the reference counter
+    called_globals_.back().second->DecRef();
+    called_globals_.pop_back();
+  }
+}
+
+inline void CallGraphEntry::AddCalledGlobal(CallGraphEntry* cg_node) {
+  called_globals_.emplace_back(global_, cg_node);
+  // Increment the reference to indicate that another call site is found for
+  // the callee in `cg_node`.
+  cg_node->IncRef();
+  // Mark the global function as recursive if it calls itself.
+  if (global_ == cg_node->GetGlobalVar()) {
+    cg_node->is_recursive_ = true;
+  }
+}
+
+// Remove an edge from the current global function to the callee.
+void CallGraphEntry::RemoveCallTo(const GlobalVar& callee) {
```

Comment:
no one uses this function?

Reply:
We will use it in the inline pass to update the graph.